idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,851 @@
+            footer="=",
+        )
+
+    def _check_unit_consistency(self):
+        # Check unit consistency
+        # TODO: replace once Pyomo method ready
+        # return identify_inconsistent_units(self.model)
+        inconsistent_units = ComponentSet()
+        for o in self.model.component_data_objects(
+            [Constraint, Expression, Objective], descend_into=True
+        ):
+            try:
+                assert_units_consistent(o)
+            except UnitsError:
+                inconsistent_units.add(o)
+        return inconsistent_units
+
+    def display_components_with_inconsistent_units(self, stream=stdout):
+        """
+        Prints a list of all Constraints, Expressions and Objectives in the
+        model with inconsistent units of measurement.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        self._write_report_section(
+            stream=stream,
+            lines_list=self._check_unit_consistency(),
+            title=f"The following component(s) have unit consistency issues:",
+            end_line="For more details on constraint violations, import the "
+            "assert_units_consistent method\nfrom pyomo.util.check_units",
+            header="=",
+            footer="=",
+        )
+
+    def display_constraints_with_large_residuals(self, stream=stdout):
+        """
+        Prints a list of Constraints with residuals greater than a specified tolerance.
+        Tolerance can be set in the class configuration options.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        self._write_report_section(
+            stream=stream,
+            lines_list=large_residuals_set(self.model, tol=self.residual_tolerance),
+            title=f"The following constraint(s) have large residuals:",
+            header="=",
+            footer="=",
+        )
+
+    def check_dulmage_mendelsohn_partition(self):
+        """
+        Performs a Dulmage-Mendelsohn partitioning on the model and returns
+        the over- and under-constraint sub-problems..
+
+        Returns:
+            list of variables in the under-constrained set
+            list of constraints in the under-constrained set
+            list of variables in the over-constrained set
+            list of constraints in the over-constrained set
+
+        """
+        igraph = IncidenceGraphInterface(self.model)
+        var_dm_partition, con_dm_partition = igraph.dulmage_mendelsohn()
+
+        # Collect under- and order-constrained sub-system
+        uc_var = var_dm_partition.unmatched + var_dm_partition.underconstrained
+        uc_con = con_dm_partition.underconstrained
+        oc_var = var_dm_partition.overconstrained
+        oc_con = con_dm_partition.overconstrained + con_dm_partition.unmatched
+
+        return uc_var, uc_con, oc_var, oc_con
```

Comment:
For large under/over constrained subsystems, decomposing into independent blocks (connected components) is very useful. This can be done with:
```python
uc_vblocks, uc_cblocks = igraph.get_connected_components(uc_var, uc_con)
oc_vblocks, oc_cblocks = igraph.get_connected_components(oc_var, oc_con)
```
The partition of the systems are lists-of-lists, so the overconstrained variables can be recovered with `oc_con = sum(oc_vblocks, [])`, or these independent blocks can be printed individually (which I would recommend).

Reply:
Thanks for the example - this was on my list of things to do.